runtime/permissions/lib.rs

```diff
@@ -674,13 +786,123 @@ impl<TQuery: QueryDescriptor> UnaryPermission<TQuery> {
   }
 }
 
-#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+#[derive(Clone, Debug)]
 pub struct PathQueryDescriptor {
-  pub requested: String,
-  pub resolved: PathBuf,
+  path: PathBuf,
+  /// Custom requested display name when differs from resolved.
+  requested: Option<String>,
+  is_windows_device_path: bool,
+}
+
+impl PartialEq for PathQueryDescriptor {
+  fn eq(&self, other: &Self) -> bool {
+    self.path == other.path
+  }
+}
+
+impl Eq for PathQueryDescriptor {}
+
+impl Hash for PathQueryDescriptor {
+  fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+    self.path.hash(state);
+  }
 }
 
 impl PathQueryDescriptor {
+  pub fn new(
+    sys: &impl sys_traits::EnvCurrentDir,
+    path: Cow<'_, Path>,
+  ) -> Result<Self, PathResolveError> {
+    let path_bytes = path.as_os_str().as_encoded_bytes();
+    if path_bytes.is_empty() {
+      return Err(PathResolveError::EmptyPath);
+    }
+    let is_windows_device_path = cfg!(windows)
+      && path_bytes.starts_with(br"\\.\")
+      && !path_bytes.contains(&b':');
+    let (path, requested) = if is_windows_device_path {
+      // On Windows, normalize_path doesn't work with device-prefix-style
+      // paths. We pass these through.
+      (path.into_owned(), None)
+    } else if path.is_absolute() {
+      (normalize_path(path.as_ref()), None)
+    } else {
+      let cwd = sys
+        .env_current_dir()
+        .map_err(PathResolveError::CwdResolve)?;
+      (
+        normalize_path(cwd.join(path.as_ref())),
+        Some(path.to_string_lossy().into_owned()),
+      )
+    };
+    Ok(Self {
+      path,
+      requested,
+      is_windows_device_path,
+    })
+  }
+
+  pub fn new_known_cwd(path: Cow<'_, PathBuf>, cwd: &Path) -> Self {
+    let path_bytes = path.as_os_str().as_encoded_bytes();
+    let is_windows_device_path = cfg!(windows)
+      && path_bytes.starts_with(br"\\.\")
+      && !path_bytes.contains(&b':');
+    let (path, display) = if is_windows_device_path {
+      // On Windows, normalize_path doesn't work with device-prefix-style
+      // paths. We pass these through.
+      (path.into_owned(), None)
+    } else if path.is_absolute() {
+      (normalize_path(path.as_ref()), None)
+    } else {
+      (
+        normalize_path(cwd.join(path.as_ref())),
+        Some(path.to_string_lossy().into_owned()),
+      )
+    };
+    Self {
+      path,
+      requested: display,
+      is_windows_device_path,
+    }
+  }
+
+  pub fn new_known_absolute(path: Cow<'_, PathBuf>) -> Self {
+    let path_bytes = path.as_os_str().as_encoded_bytes();
```

Comment:
Should there be a `debug_assert!` to assert the path is actually absolute?

Reply:
This is actually mostly only used in the tests and in the tests it's not always absolute 😅 (ex. `/something`, which won't be absolute on windows)

There's one place it's used in the regular code and seems ok